go/kbfs/libfs/wrapped_read_file.go

```diff
@@ -0,0 +1,130 @@
+// Copyright 2019 Keybase Inc. All rights reserved.
+// Use of this source code is governed by a BSD
+// license that can be found in the LICENSE file.
+
+package libfs
+
+import (
+	"context"
+	"io"
+	"time"
+
+	"github.com/keybase/client/go/kbfs/libkbfs"
+	"github.com/keybase/client/go/logger"
+	"github.com/pkg/errors"
+	billy "gopkg.in/src-d/go-billy.v4"
+)
+
+const (
+	// Debug tag ID for an individual FS operation.
+	ctxFSOpID = "FSID"
+)
+
+type ctxFSTagKey int
+
+const (
+	ctxFSIDKey ctxFSTagKey = iota
+)
+
+// wrappedReadFile is a read-only file that serves data from a given
+// `reader` function.
+type wrappedReadFile struct {
+	name   string
+	reader func(context.Context) ([]byte, time.Time, error)
+	log    logger.Logger
+
+	// TODO: proper locking for `nextRead` if we ever use it outside
+	// of libkbfs node-wrapping (which always uses `ReadAt`).
+	nextRead int
+}
+
+var _ billy.File = (*wrappedReadFile)(nil)
+
+func (wrf *wrappedReadFile) getDataAndTime() ([]byte, time.Time) {
+	ctx := libkbfs.CtxWithRandomIDReplayable(
+		context.Background(), ctxFSIDKey, ctxFSOpID, nil)
+	data, t, err := wrf.reader(ctx)
+	if err != nil {
+		wrf.log.CDebugf(ctx, "Couldn't read wrapped file: %+v", err)
+		return nil, time.Time{}
+	}
+	return data, t
+}
+
+func (wrf *wrappedReadFile) Len() int {
+	data, _ := wrf.getDataAndTime()
+	return len(data)
+}
+
+func (wrf *wrappedReadFile) Name() string {
+	return wrf.name
+}
+
+func (wrf *wrappedReadFile) Write(_ []byte) (n int, err error) {
+	return 0, errors.New("wrapped read files can't be written")
+}
+
+func (wrf *wrappedReadFile) Read(p []byte) (n int, err error) {
+	data, _ := wrf.getDataAndTime()
+
+	if wrf.nextRead >= len(data) {
+		return 0, io.EOF
+	}
+	n = copy(p, data[wrf.nextRead:])
```

Comment:
Doesn't this risk that `getDataAndTime` could change between calls to `Read`?

Same comment re: `ReadAt`, `Seek`, `Len`, and `GetInfo`.

Reply:
Yes, but so could any file you're reading on a file system (when it's modified by some other process).  (And the previous libfuse-based code had the same issue, where it read the file once on a stat and then again on each read, so I don't think this introduces anything new.)

This is also the reason we dial the kernel caching duration to 0 for all of this data.

Reply:
OK fair enough.